src/main/java/hudson/remoting/FileSystemJarCache.java

```diff
@@ -71,8 +71,13 @@ protected URL retrieve(Channel channel, long sum1, long sum2) throws IOException
                 tmp.renameTo(target);
 
                 if (target.exists()) {
+                    Checksum expected = new Checksum(sum1, sum2);
+                    Checksum actual = Checksum.forFile(target);
+                    if (!expected.equals(actual)) {
+                        throw new IOException("Checksum of retrieved jar doesn't match: " + target);
```

Comment:
I would suggest including the two checksums in the detail message, for possible future diagnosis.